src/buoyancy.jl

```diff
@@ -0,0 +1,194 @@
+
+Returns parameters for an idealized nonlinear equation of state with reference density 
+`ρ₀`. The `flavor` of the nonlinear equation of state is a symbol corresponding to one of 
+the 5 equations of state in equations (12)--(15) of Roquet et al., "Defining a Simplified 
+yet 'Realistic' Equation of State for Seawater", Journal of Physical Oceanography (2015). 
+The equations of state define the density anomaly `ρ′`, and have the polynomial form
+
+    `ρ′(T, S, D) = Σᵢⱼₐ Rᵢⱼₐ Tⁱ Sʲ Dᵃ`,
+
+where `T` is conservative temperature, `S` is absolute salinity, and `D` is the 
+geopotential depth, currently just `D = -z`. The `Rᵢⱼₐ` are constant coefficients.
+
+Flavors of idealized nonlinear equations of state
+=================================================
+
+    - `:linear`: a linear equation of state, `ρ′ = R₁₀₀ * T + R₀₁₀ * S`
+
+    - `:cabbeling`: includes quadratic temperature term, 
+                    `ρ′ = R₁₀₀ * T + R₀₁₀ * S + R₀₂₀ * T^2`
+
+    - `:cabbeling_thermobaricity`: includes 'thermobaricity' term, 
+                                   `ρ′ = R₁₀₀ * T + R₀₁₀ * S + R₀₂₀ * T^2 + R₀₁₁ * T * D`
+
+    - `:freezing`: same as `:cabbeling_thermobaricity` with modified constants to increase 
+                   accuracy near freezing
+
+    - `:second_order`: includes quadratic salinity, halibaricity, and thermohaline term,
+                       `ρ′ = R₁₀₀ * T + R₀₁₀ * S + R₀₂₀ * T^2 + R₀₁₁ * T * D`
+                             + R₂₀₀ * S^2 + R₁₀₁ * S * D + R₁₁₀ * S * T`
+"""
+function RoquetIdealizedNonlinearEOS(T, flavor=:cabbeling_thermobaricity; coeffs=roquet_coeffs[flavor], ρ₀=1025)
+    typed_coeffs = type_convert_roquet_coeffs(coeffs)
+    return RoquetIdealizedNonlinearEOS{flavor, typeof(typed_coeffs), T}(ρ₀, typed_coeffs)
+end
+
+RoquetIdealizedNonlinearEOS(flavor; kwargs...) = RoquetIdealizedNonlinearEOS(Float64, flavor; kwargs...)
+
+@inline ρ′(i, j, k, eos::RoquetIdealizedNonlinearEOS, C) = 
+    @inbounds (   eos.coeffs.R₁₀₀ * C.S[i, j, k]
+                + eos.coeffs.R₀₁₀ * C.T[i, j, k]
+                + eos.coeffs.R₀₂₀ * C.T[i, j, k]^2
+                + eos.coeffs.R₀₁₁ * C.T[i, j, k] * D(i, j, k, grid)
+                + eos.coeffs.R₂₀₀ * C.S[i, j, k]^2
+                + eos.coeffs.R₁₀₁ * C.S[i, j, k] * D(i, j, k, grid)
+                + eos.coeffs.R₁₁₀ * C.S[i, j, k] * C.T[i, j, k] )
+
+@inline thermal_expansion(i, j, k, eos::RoquetIdealizedNonlinearEOS, C) = 
+    @inbounds (   eos.coeffs.R₀₁₀
+                + 2 * eos.coeffs.R₀₂₀ * C.T[i, j, k]
+                + eos.coeffs.R₀₁₁ * D(i, j, k, grid)
+                + eos.coeffs.R₁₁₀ * C.T[i, j, k] )
+
+@inline haline_contraction(i, j, k, eos::RoquetIdealizedNonlinearEOS, C) = 
+    @inbounds (   eos.coeffs.R₁₀₀
+                + 2 * eos.coeffs.R₂₀₀ * C.S[i, j, k]
+                + eos.coeffs.R₁₀₁ * D(i, j, k, grid)
+                + eos.coeffs.R₁₁₀ * C.S[i, j, k] 
+
+#=
+# Optimiziations: do we need?
```

Comment:
Hmmm, I'd like to think that the compiler/LLVM will optimize away any `+ 0` terms as `RoquetIdealizedNonlinearEOS` is a concrete type so it should know what's a zero at compile time.

Seems like it might be good to just keep the general case for now as this won't be a widely used feature for now, so less bugs and less code to maintain. It's also a pretty minor optimization.